platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/testapp/camera/CameraTest.java

```diff
@@ -0,0 +1,301 @@
+import com.mapbox.mapboxsdk.geometry.LatLngBounds;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+import com.mapbox.mapboxsdk.testapp.activity.BaseTest;
+import com.mapbox.mapboxsdk.testapp.activity.espresso.DeviceIndependentTestActivity;
+import com.mapbox.mapboxsdk.testapp.utils.TestConstants;
+
+import org.junit.Test;
+
+import static com.mapbox.mapboxsdk.testapp.action.MapboxMapAction.invoke;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+public abstract class CameraTest extends BaseTest {
+
+  private final CountingIdlingResource animationIdlingResource =
+    new CountingIdlingResource("animation_idling_resource");
+
+  @Override
+  protected Class getActivityClass() {
+    return DeviceIndependentTestActivity.class;
+  }
+
+  @Override
+  public void beforeTest() {
+    super.beforeTest();
+    IdlingRegistry.getInstance().register(animationIdlingResource);
+  }
+
+  @Test
+  public void testToCameraPositionTarget() {
+    validateTestSetup();
+    invoke(mapboxMap, (uiController, mapboxMap) -> {
+      float zoom = 1.0f;
+      LatLng moveTarget = new LatLng(1, 1);
+      CameraPosition initialPosition = new CameraPosition.Builder().target(
+        new LatLng()).zoom(zoom).bearing(0).tilt(0).padding(new double[4]).build();
+      CameraPosition cameraPosition = mapboxMap.getCameraPosition();
+      assertEquals("Default camera position should match default", cameraPosition, initialPosition);
+
+      animationIdlingResource.increment();
+      executeCameraMovement(CameraUpdateFactory.newLatLng(moveTarget), new MapboxMap.CancelableCallback() {
+        @Override
+        public void onCancel() {
+          verifyCameraPosition(mapboxMap, moveTarget, zoom, 0, 0, new double[4]);
+          animationIdlingResource.decrement();
+        }
+
+        @Override
+        public void onFinish() {
+          verifyCameraPosition(mapboxMap, moveTarget, zoom, 0, 0, new double[4]);
+          animationIdlingResource.decrement();
+        }
+      });
+    });
+
+    Espresso.onIdle();
+  }
+
+  @Test
+  public void testToCameraPositionTargetZoom() {
+    validateTestSetup();
+    invoke(mapboxMap, (uiController, mapboxMap) -> {
+      final float moveZoom = 15.5f;
+      final LatLng moveTarget = new LatLng(1.0000000001, 1.0000000003);
+
+      animationIdlingResource.increment();
+      executeCameraMovement(CameraUpdateFactory.newLatLngZoom(moveTarget, moveZoom),
+        new MapboxMap.CancelableCallback() {
+          @Override
+          public void onCancel() {
+            verifyCameraPosition(mapboxMap, moveTarget, moveZoom, 0, 0, new double[4]);
+            animationIdlingResource.decrement();
+          }
+
+          @Override
+          public void onFinish() {
+            verifyCameraPosition(mapboxMap, moveTarget, moveZoom, 0, 0, new double[4]);
+            animationIdlingResource.decrement();
+          }
+        });
+    });
+
+    Espresso.onIdle();
```

Comment:
is there a hard requirement on using espresso here? Since tests have been rewritten, I feel it the same as introducing new tests. Espresso should only be used if necessary.

Reply:
Only the abstraction has been introduced, but are you suggesting unit tests? This wouldn't allow us to test JNI conversion code.

Reply:
I'm not seeing the need for using Espresso in this kind of test. The UiSelector setup with Espresso.onIdle() has proven to be unreliable in comparison with a latch setup.

Reply:
Address in https://github.com/mapbox/mapbox-gl-native/pull/15444/commits/a00fc3aadd834815deef4165073da8813ceb8f59.